cnudie/upload.py

```diff
@@ -0,0 +1,135 @@
+import enum
+import hashlib
+
+import ccc.oci
+import gci.componentmodel as cm
+import gci.oci
+import json
+import oci.model as om
+import dataclasses
+
+
+class UploadMode(enum.Enum):
+    SKIP = 'skip'
+    FAIL = 'fail'
+    OVERWRITE = 'overwrite'
+
+
+def _normalise_component_name(component_name: str) -> str:
+    return component_name.lower()  # oci-spec allows only lowercase
+
+
+def _target_oci_ref(
```

Comment:
ditto - considering this is also used in replicate.py / not specific to uploading, but also required for downloading